File: Crio/crio_programming_interview_problems-master/MaxSubmatrix/MaxSubmatrix.py
def maxSubmatrix(n,m,matrix):
    # CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    max_sum = -1e9
    for row in matrix:
        for x in row:
            max_sum = max(max_sum,x)
    
    if max_sum<0:
        return max_sum

    
    for left in range(m):
        tmp = [0 for _ in range(n)]
        for right in range(left,m):
            for i in range(n):
                tmp[i]+=matrix[i][right]
            curr_sum=0
            for i in range(n):
                curr_sum+=tmp[i]
                if(curr_sum<0):
                    curr_sum=0
            max_sum = max(max_sum,curr_sum)
    
    

    return max_sum
    # CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

def main():
    n,m = map(int,input().strip().split())
    matrix = []
    for _ in range(n):
        tmp = list(map(int,input().strip().split()))
        matrix.append(tmp)  
    ans = maxSubmatrix(n,m,matrix)
    print(ans)

if __name__=="__main__":
    main()

File: Crio/crio_programming_interview_problems-master/MaxSubmatrix/MaxSubmatrix.java
import java.util.*;
class MaxSubmatrix {
	public static long maxSubmatrix(int [][]arr, int n, int m) {
		// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
		long ans = (long) - 1e9;
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				ans = Math.max(ans, arr[i][j]);
			}
		}
		if (ans < 0) return ans;
		for (int left = 0; left < m; left++) {
			long []maxCol = new long[n];
			for ( int right = left; right < m; right++) {
				for ( int k = 0; k < n; k++) {
					maxCol[k] = maxCol[k] + arr[k][right];
				}
				long temp = 0;
				for (int i = 0; i < n; i++) {
					temp += maxCol[i];
					if (temp < 0) temp = 0;
				}
				ans = Math.max(ans, temp);
			}
		}
		return ans;
		// CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
	}
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		int m = sc.nextInt();
		int [][]arr = new int[n][m];
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				arr[i][j] = sc.nextInt();
			}
		}
		long ans = maxSubmatrix(arr, n, m);
		System.out.println(ans);
	}
}